File: BEmu/ReferenceDataRequest/RequestReference.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="ReferenceDataRequest/RequestReference.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class RequestReference : Request
    {
        private readonly RequestReferenceElementStringArray _securities, _fields;
        private readonly RequestReferenceElementOverrideArray _overrides;

        public RequestReference()
        {
            this._securities = new RequestReferenceElementStringArray("securities");
            this._fields = new RequestReferenceElementStringArray("fields");
            this._overrides = new RequestReferenceElementOverrideArray();
        }

        internal List<string> Securities { get { return this._securities.Values; } }
        internal List<string> Fields { get { return this._fields.Values; } }

        public override IEnumerable<Element> Elements
        {
            get
            {
                yield return this._securities;
                yield return this._fields;

                if (this._overrides.NumValues > 0)
                    yield return this._overrides;
            }
        }

        public override bool HasElement(string name)
        {
            string strName = name.ToUpper();

            if (this._securities.Name.ToString().ToUpper() == strName)
                return this._securities.Values.Any();
            else if (this._fields.Name.ToString().ToUpper() == strName)
                return this._fields.Values.Any();
            else if (this._overrides.Name.ToString().ToUpper() == strName)
                return this._overrides.NumValues > 0;
            else
                return false;
        }

        public override Element this[string elementName]
        {
            get
            {
                string strName = elementName.ToUpper();

                if (this._securities.Name.ToString().ToUpper() == strName)
                    return this._securities;
                else if (this._fields.Name.ToString().ToUpper() == strName)
                    return this._fields;
                else if (this._overrides.Name.ToString().ToUpper() == strName)
                    return this._overrides;
                else
                    throw new NotImplementedException("BEmu.ReferenceDataRequest.RequestReference: public Element this[string elementName] not supported");
            }
        }

        public override void Append(string name, string elementValue)
        {
            switch (name.ToLower())
            {
                case "securities":
                    this._securities.AddValue(elementValue);
                    break;
                case "fields":
                    this._fields.AddValue(elementValue);
                    break;
                default:
                    throw new ArgumentException(string.Format("BEmu.RequestReference.Append: Element name {0} not supported", name));
            }
        }

        public override string ToString()
        {
            StringBuilder result = new StringBuilder();
            result.AppendFormat("ReferenceDataRequest = {{{0}", Environment.NewLine);

            if (this._securities.NumValues > 0)
                result.Append(this._securities.PrettyPrint(1));

            if (this._fields.NumValues > 0)
                result.Append(this._fields.PrettyPrint(1));

            if (this._overrides.NumValues > 0)
                result.Append(this._overrides.PrettyPrint(1));

            result.Append("}");

            return result.ToString();
        }

    }
}

File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/ReferenceDataRequest/RequestReference.java
package com.bemu.BEmu.ReferenceDataRequest;

import com.bemu.BEmu.Request;
import com.bemu.BEmu.Element;
import java.util.List;

public class RequestReference extends Request
{
    private final RequestReferenceElementStringArray _securities, _fields;
    private final RequestReferenceElementOverrideArray _overrides;

    public RequestReference()
    {
        this._securities = new RequestReferenceElementStringArray("securities");
        this._fields = new RequestReferenceElementStringArray("fields");
        this._overrides = new RequestReferenceElementOverrideArray();
    }
    
    List<String> securities()
    {
    	return this._securities.values();
    }
    
    List<String> fields()
    {
    	return this._fields.values();
    }
    
    public Element getElement(String name) throws Exception
    {
        String strName = name.toUpperCase();

        if (this._securities.name().toString().toUpperCase().equals(strName))
            return this._securities;
        
        else if (this._fields.name().toString().toUpperCase().equals(strName))
            return this._fields;
        
        else if (this._overrides.name().toString().toUpperCase().equals(strName))
            return this._overrides;
            
        else
            return super.getElement(name);
    }
    
    public boolean hasElement(String name)
    {
        String strName = name.toUpperCase();

        if (this._securities.name().toString().toUpperCase().equals(strName))
            return this._securities.values().size() > 0;
        
        else if (this._fields.name().toString().toUpperCase().equals(strName))
            return this._fields.values().size() > 0;
        
        else if (this._overrides.name().toString().toUpperCase().equals(strName))
            return this._overrides.numValues() > 0;
            
        else
            return false;
    }
    
    public void append(String name, String elementValue) throws Exception
    {
    	String lower = name.toLowerCase();
    	
    	if(lower.equals("securities"))
    		this._securities.addValue(elementValue);
    	
    	else if(lower.equals("fields"))
    		this._fields.addValue(elementValue);
    	
    	else
    		throw new Exception(String.format("BEmu.RequestReference.Append: Element name %s not supported", name));
    }
    
    public String toString()
    {
        StringBuilder result = new StringBuilder();
        result.append(String.format("ReferenceDataRequest = {%s", System.getProperty("line.separator")));

        if (this._securities.numValues() > 0)
            result.append(this._securities.prettyPrint(1));

        if (this._fields.numValues() > 0)
            result.append(this._fields.prettyPrint(1));

        if (this._overrides.numValues() > 0)
            result.append(this._overrides.prettyPrint(1));

        result.append("}");

        return result.toString();
    }
    
}
